featurebyte/routes/deployment/controller.py

```diff
@@ -112,19 +119,72 @@ async def update_deployment(
+        DocumentUpdateError
+            If there is an existing task to update deployment that is conflicting with the current request.
+            If the task submission lock is not acquired.
         """
         # check if deployment exists
         deployment = await self.service.get_document(document_id=document_id)
-        if data.enabled is not None and data.enabled != deployment.enabled:
-            payload = DeploymentCreateUpdateTaskPayload(
-                deployment_payload=UpdateDeploymentPayload(enabled=data.enabled),
-                user_id=self.service.user.id,
-                catalog_id=self.service.catalog_id,
-                output_document_id=document_id,
-            )
-            task_id = await self.task_controller.task_manager.submit(payload=payload)
-            return await self.task_controller.get_task(task_id=str(task_id))
-        return None
+
+        # create a lock that maximum live for REDIS_LOCK_EXPIRE_TIME seconds
+        redis_conn = Redis.from_url(REDIS_URI)
+        lock = Lock(
+            redis_conn, f"deployment:{deployment.id}:update", timeout=REDIS_LOCK_EXPIRE_TIME
+        )
+
+        # set blocking to False to prevent waiting for the lock
+        if lock.acquire(blocking=False):
+            task_manager = self.task_controller.task_manager
+            try:
+                # check whether there are existing task
+                tasks, total = await task_manager.list_tasks(
+                    query_filter={
+                        "kwargs.command": "DEPLOYMENT_CREATE_UPDATE",
+                        "kwargs.output_document_id": str(document_id),
+                        "status": {"$in": ["STARTED", "PENDING"]},
+                    }
+                )
+                if total > 0:
+                    task = tasks[0]
+                    task_enabled = task.payload["deployment_payload"]["enabled"]
+                    if task_enabled == data.enabled:
+                        return task
+                    raise DocumentUpdateError(
+                        f"There is an existing task to update deployment (enabled: {task_enabled})."
+                    )
+
+                if data.enabled is None or data.enabled == deployment.enabled:
+                    return None
+
+                payload = DeploymentCreateUpdateTaskPayload(
+                    deployment_payload=UpdateDeploymentPayload(enabled=data.enabled),
+                    user_id=self.service.user.id,
+                    catalog_id=self.service.catalog_id,
+                    output_document_id=document_id,
+                )
+                task_id = await task_manager.submit(payload=payload)
+
+                # check whether the mongo is updated, if not, wait for the mongo to be updated
+                start = time.time()
+                while time.time() - start < MAX_MONGO_TASK_UPDATE_WAIT_TIME:
+                    _, total = await task_manager.list_tasks(query_filter={"_id": str(task_id)})
```

Comment:
When will a task show up in `list_tasks` after it is submitted? Is it when it started running?